### ISSUE TITLE
eval in package DB can't see caller's lexicals in certain cases

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16230]
Migrated from [rt.perl.org#132414](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132414) (status was 'open')

Searchable as RT132414$

**Comment by p5pRT** [comment]
### From @jimav

This is a bug report for perl from jim\.avera@&#8203;gmail\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.24\.1\.



---
I have a debugging library which uses eval in package DB to
access variables in the caller's context \(per perldoc \-f eval\)\.

However lexicals are sometimes not accessible\, and it seems to
depend on "irrelevant" details such as whether some not\-yet\-executed
code exists following the call in the same block\.   I don't know if this 
is a regression\,
but I noticed it shortly after upgrading to perl 5\.24\.1

In the example below\, if there is any statement following
the call into package DB\, then it works as expected\, i\.e\.\, the
lexicals $key and $value are visible\.  But as shown the
code dies with an eval error\.

Can someone shed light on this?
Is there anything my code inside the DB function can do to avoid this 
problem?

\#\!/usr/bin/perl
use strict; use warnings;

package DB;

sub my\_interpolate\($\) \{
   my\($s\) = @&#8203;\_;
   my $result = eval "$s"; die "eval '$s' Failed&#8203;: $@&#8203;" if $@&#8203;;
   print "eval '$s' => $result\\n";
\}

package main;

my %hash = \(A => 100\, B => 200\);

while \(my \($key\, $value\) = each %hash\) \{
   \# \*\*DIES HERE\*\* with
   \#    eval '$key' Failed&#8203;: Global symbol "$key" requires explicit 
package name\.\.\.
   DB&#8203;::my\_interpolate '$key';

   \#1;    \# un\-comment this and the problem vanishes\!
\}



---


---
Flags&#8203;:
     category=core
     severity=low


---
Site configuration information for perl 5\.24\.1&#8203;:

Configured by Debian Project at Sat Mar 18 17&#8203;:00&#8203;:39 UTC 2017\.

Summary of my perl5 \(revision 5 version 24 subversion 1\) configuration&#8203;:

   Platform&#8203;:
     osname=linux\, osvers=3\.16\.0\, archname=x86\_64\-linux\-gnu\-thread\-multi
     uname='linux localhost 3\.16\.0 \#1 smp debian 3\.16\.0 x86\_64 gnulinux '
     config\_args='\-Dusethreads \-Duselargefiles \-Dcc=x86\_64\-linux\-gnu\-gcc 
\-Dcpp=x86\_64\-linux\-gnu\-cpp \-Dld=x86\_64\-linux\-gnu\-gcc \-Dccflags=\-DDEBIAN 
\-Wdate\-time \-D\_FORTIFY\_SOURCE=2 \-g \-O2 
\-fdebug\-prefix\-map=/build/perl\-gaNj0t/perl\-5\.24\.1=\. 
\-fstack\-protector\-strong \-Wformat \-Werror=format\-security \-Dldflags= 
\-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared 
\-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC 
\-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.24 
\-Darchlib=/usr/lib/x86\_64\-linux\-gnu/perl/5\.24 \-Dvendorprefix=/usr 
\-Dvendorlib=/usr/share/perl5 
\-Dvendorarch=/usr/lib/x86\_64\-linux\-gnu/perl5/5\.24 
\-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.24\.1 
\-Dsitearch=/usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1 
\-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 
\-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3
\-Dusesitecustomize \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl 
\-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio 
\-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-dEs 
\-Duseshrplib \-Dlibperl=libperl\.so\.5\.24\.1'
     hint=recommended\, useposix=true\, d\_sigaction=define
     useithreads=define\, usemultiplicity=define
     use64bitint=define\, use64bitall=define\, uselongdouble=undef
     usemymalloc=n\, bincompat5005=undef
   Compiler&#8203;:
     cc='x86\_64\-linux\-gnu\-gcc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE 
\-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include 
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
     optimize='\-O2 \-g'\,
     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv 
\-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
     ccversion=''\, gccversion='6\.3\.0 20170316'\, gccosandvers=''
     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\, 
doublekind=3
     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, 
longdblsize=16\, longdblkind=3
     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
     alignbytes=8\, prototype=define
   Linker and Libraries&#8203;:
     ld='x86\_64\-linux\-gnu\-gcc'\, ldflags =' \-fstack\-protector\-strong 
\-L/usr/local/lib'
     libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/6/include\-fixed 
/usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib 
/usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
     libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
     perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
     libc=libc\-2\.24\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.24
     gnulibc\_version='2\.24'
   Dynamic Linking&#8203;:
     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
     cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib 
\-fstack\-protector\-strong'

Locally applied patches&#8203;:
     DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS 
default for modules installed from CPAN\.
     DEBPKG&#8203;:debian/db\_file\_ver \- https://bugs.debian.org/340047 Remove 
overly restrictive DB\_File version check\.
     DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with 
Debian\-specific information\.
     DEBPKG&#8203;:debian/enc2xs\_inc \- https://bugs.debian.org/290336 Tweak 
enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
     DEBPKG&#8203;:debian/errno\_ver \- https://bugs.debian.org/343351 Remove 
Errno version check due to upgrade problems with long\-running processes\.
     DEBPKG&#8203;:debian/libperl\_embed\_doc \- https://bugs.debian.org/186778 
Note that libperl\-dev package is required for embedded linking
     DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
     DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site 
install directories
     DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: set location of 
libperl\.a under /usr/lib
     DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or 
perllocal\.pod for perl or vendor
     DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the 
binary targets\.
     DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install 
\.packlist files for core or vendor\.
     DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH 
as per Debian policy\.
     DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to 
/etc/perl/Net as /usr may not be writable\.
     DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
     DEBPKG&#8203;:debian/prune\_libs \- https://bugs.debian.org/128355 Prune the 
list of libraries wanted to what we actually need\.
     DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] 
https://bugs.debian.org/100195 Document the Net&#8203;::SMTP 'Port' option
     DEBPKG&#8203;:debian/perlivp \- https://bugs.debian.org/510895 Make perlivp 
skip include directories in /usr/local
     DEBPKG&#8203;:debian/deprecate\-with\-apt \- https://bugs.debian.org/747628 
Point users to Debian packages of deprecated core modules
     DEBPKG&#8203;:debian/squelch\-locale\-warnings \- 
https://bugs.debian.org/508764 Squelch locale warnings in Debian package 
maintainer scripts
     DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the 
upstream Git repository
     DEBPKG&#8203;:debian/patchlevel \- https://bugs.debian.org/567489 List 
packaged patches for 5\.24\.1\-2ubuntu1 in patchlevel\.h
     DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- https://bugs.debian.org/628493 
\[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
     DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- 
https://bugs.debian.org/628522 \[rt\.cpan\.org \#68613\] Document that 
CCFLAGS should include $Config\{ccflags\}
     DEBPKG&#8203;:debian/find\_html2text \- https://bugs.debian.org/640479 
Configure CPAN&#8203;::Distribution with correct name of html2text
     DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- 
https://bugs.debian.org/668490 Invoke x\-terminal\-emulator rather than 
xterm in perl5db\.pl
     DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- 
https://bugs.debian.org/688842 Fix CPAN&#8203;::FirstTime defaults with 
nonexisting site dirs if a parent is writable
     DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] 
https://bugs.debian.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' 
option not respected
     DEBPKG&#8203;:debian/regen\-skip \- Skip a regeneration check in unrelated 
git repositories
     DEBPKG&#8203;:debian/makemaker\-pasthru \- https://bugs.debian.org/758471 
Pass LD settings through to subdirectories
     DEBPKG&#8203;:debian/makemaker\-manext \- https://bugs.debian.org/247370 
Make EU&#8203;::MakeMaker honour MANnEXT settings in generated manpage headers
     DEBPKG&#8203;:debian/devel\-ppport\-reproducibility \- 
https://bugs.debian.org/801523 Sort the list of XS code files when 
generating RealPPPort\.xs
     DEBPKG&#8203;:debian/encode\-unicode\-bom\-doc \- 
https://bugs.debian.org/798727 Document Debian backport of 
Encode&#8203;::Unicode fix
     DEBPKG&#8203;:debian/kfreebsd\-softupdates \- https://bugs.debian.org/796798 
Work around Debian Bug\#796798
     DEBPKG&#8203;:fixes/autodie\-scope \- https://bugs.debian.org/798096 Fix a 
scoping issue with "no autodie" and the "system" sub
     DEBPKG&#8203;:fixes/crosscompile\-no\-targethost \- \[23695c0\] \[perl \#127234\] 
Fix the Configure escape with usecrosscompile but no targethost
     DEBPKG&#8203;:fixes/memoize\-pod \- \[rt\.cpan\.org \#89441\] Fix POD errors in 
Memoize
     DEBPKG&#8203;:fixes/ok\-pod \- Added encoding for pod\.
     DEBPKG&#8203;:debian/hurd\-softupdates \- https://bugs.debian.org/822735 Fix 
t/op/stat\.t failures on hurd
     DEBPKG&#8203;:fixes/nntp\_docs \- https://bugs.debian.org/51962 Net&#8203;::NNTP&#8203;: 
Correct innd/nnrpd confusion in relation to Reader option
     DEBPKG&#8203;:fixes/math\_complex\_doc\_great\_circle \- 
https://bugs.debian.org/697567 \[rt\.cpan\.org \#114104\] Math&#8203;::Trig&#8203;: clarify 
definition of great\_circle\_midpoint
     DEBPKG&#8203;:fixes/math\_complex\_doc\_see\_also \- 
https://bugs.debian.org/697568 \[rt\.cpan\.org \#114105\] Math&#8203;::Trig&#8203;: add 
missing SEE ALSO
     DEBPKG&#8203;:fixes/math\_complex\_doc\_angle\_units \- 
https://bugs.debian.org/731505 \[rt\.cpan\.org \#114106\] Math&#8203;::Trig&#8203;: 
document angle units
     DEBPKG&#8203;:fixes/cpan\_web\_link \- https://bugs.debian.org/367291 CPAN&#8203;: 
Add link to main CPAN web site
     DEBPKG&#8203;:fixes/time\_piece\_doc \- https://bugs.debian.org/817925 
Time&#8203;::Piece&#8203;: Improve documentation for add\_months and add\_years
     DEBPKG&#8203;:fixes/perlbug\-refactor \- https://bugs.debian.org/822463 
\[perl \#128020\] perlbug&#8203;: Refactor duplicated file reading code
     DEBPKG&#8203;:fixes/perlbug\-linewrap \- https://bugs.debian.org/822463 
\[perl \#128020\] perlbug&#8203;: wrap overly long lines
     DEBPKG&#8203;:fixes/hurd\_sigaction \- https://bugs.debian.org/825016 
\[d54f4ed\] ext/POSIX/t/sigaction\.t&#8203;: Skip uid and pid tests on GNU/Hurd
     DEBPKG&#8203;:fixes/hurd\_hints \- \[4694301\] https://bugs.debian.org/825020 
\[perl \#128279\] Modify hints for Hurd per Debian ticket 825020\.
     DEBPKG&#8203;:fixes/extutils\-parsexs\-reproducibility \- \[perl \#128517\] 
https://bugs.debian.org/829296 Make the output of ExtUtils&#8203;::ParseXS 
reproducible
     DEBPKG&#8203;:debian/CVE\-2016\-1238/sitecustomize\-in\-etc \- Look for 
sitecustomize\.pl in /etc/perl rather than sitelib on Debian systems
     DEBPKG&#8203;:debian/CVE\-2016\-1238/test\-suite\-without\-dot \- \[perl \#127810\] 
Patch unit tests to explicitly insert "\." into @&#8203;INC when needed\.
     DEBPKG&#8203;:debian/CVE\-2016\-1238/eumm\-without\-dot \- \[perl \#127810\] Add 
PERL\_USE\_UNSAFE\_INC support to EU&#8203;::MM for fortify\_inc support\.
     DEBPKG&#8203;:debian/CVE\-2016\-1238/cpan\-without\-dot \- \[perl \#127810\] Set 
PERL\_USE\_UNSAFE\_INC for cpan usage
     DEBPKG&#8203;:debian/document\_inc\_removal \- Document in perlvar that we 
remove '\.' from @&#8203;INC by default
     DEBPKG&#8203;:fixes/extutils\_makemaker\_reproducible \- 
https&#8203;://bugs\.debian\.org/835815 https://bugs.debian.org/834190 Make 
perllocal\.pod files reproducible
     DEBPKG&#8203;:debian/CVE\-2016\-1238/remove\-inc\-test \- Remove test for '\.' 
in @&#8203;INC as it might not be
     DEBPKG&#8203;:debian/customized \- Update customized\.dat for files patched 
in Debian
     DEBPKG&#8203;:fixes/file\_path\_hurd\_errno \- File\-Path&#8203;: Fix test failure in 
Hurd due to hard\-coded ENOENT
     DEBPKG&#8203;:debian/hppa\_op\_optimize\_workaround \- 
https://bugs.debian.org/838613 Temporarily lower the optimization of 
op\.c on hppa due to gcc\-6 problems
     DEBPKG&#8203;:fixes/test\-builder\-warning \- https://bugs.debian.org/840968 
Silence a 'used only once' warning in Test&#8203;::Builder
     DEBPKG&#8203;:fixes/longdblinf\-randomness \- \[dd68853\] \[perl \#130133\] 
https://bugs.debian.org/844752 Configure&#8203;: fix garbage filtering with 
80\-bit long doubles
     DEBPKG&#8203;:debian/installman\-utf8 \- https://bugs.debian.org/840211 
Generate man pages with UTF\-8 characters
     DEBPKG&#8203;:fixes/list\_assign\_leak \- \[1050723\] \[perl \#130766\] 
https://bugs.debian.org/855064 avoid a leak in list assign from/to magic 
values
     DEBPKG&#8203;:fixes/perlfunc\_inc\_doc \- \[a03e9f8\] 
https://bugs.debian.org/839536 \[perl \#130832\] Documentation fixes for 
'\.' possibly no longer being in @&#8203;INC
DEBPKG&#8203;:fixes/Compress\-Raw\-Zlib\-2\.071\-Adapt\-tests\-to\-zlib\-1\.2\.11\.patch \- 
\[PATCH\] Adapt tests to zlib 1\.2\.11
     DEBPKG&#8203;:fixes/0001\-Adapt\-tests\-to\-zlib\-1\.2\.11\.patch \- \[PATCH\] Adapt 
tests to zlib 1\.2\.11



---
@&#8203;INC for perl 5\.24\.1&#8203;:
/home/jima/perl5/lib/perl5/5\.24\.1/x86\_64\-linux\-gnu\-thread\-multi
     /home/jima/perl5/lib/perl5/5\.24\.1
     /home/jima/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
     /home/jima/perl5/lib/perl5
     /home/jima/lib/perl
     /etc/perl
     /usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1
     /usr/local/share/perl/5\.24\.1
     /usr/lib/x86\_64\-linux\-gnu/perl5/5\.24
     /usr/share/perl5
     /usr/lib/x86\_64\-linux\-gnu/perl/5\.24
     /usr/share/perl/5\.24
     /usr/local/lib/site\_perl
     /usr/lib/x86\_64\-linux\-gnu/perl\-base



---
Environment for perl 5\.24\.1&#8203;:
     HOME=/home/jima
     LANG=en\_US\.UTF\-8
     LANGUAGE=en\_US
     LC\_COLLATE=C
     LD\_LIBRARY\_PATH \(unset\)
     LOGDIR \(unset\)
PATH=/home/jima/perl5/bin&#8203;:/home/jima/bin&#8203;:/home/jima/jima\_tools/x86\_64/bin&#8203;:/home/jima/jima\_tools/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/bin/X11&#8203;:/usr/local/bin&#8203;:/usr/local/sbin&#8203;:/usr/games&#8203;:/usr/local/games&#8203;:/usr/lib/jvm/java\-8\-oracle/bin&#8203;:/usr/lib/jvm/java\-8\-oracle/db/bin&#8203;:/usr/lib/jvm/java\-8\-oracle/jre/bin&#8203;:\.
     PERL5LIB=/home/jima/perl5/lib/perl5&#8203;:/home/jima/lib/perl
     PERL\_BADLANG \(unset\)
     PERL\_LOCAL\_LIB\_ROOT=/home/jima/perl5
     PERL\_MB\_OPT=\-\-install\_base "/home/jima/perl5"
     PERL\_MM\_OPT=INSTALL\_BASE=/home/jima/perl5
     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From zefram@fysh.org

>I don't know if this 
>is a regression\,
>but I noticed it shortly after upgrading to perl 5\.24\.1

It's a regression in 5\.13\.7\.  Bisect fingers commit
eae48c8938e50ebb341a72c2886c5ae8587092a5 "refactor and regularise
label/statement grammar"\.  That commit was not intended to change this
eval behaviour\.

>Is there anything my code inside the DB function can do to avoid this 
>problem?

No\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

This issue is about nextstate ops being elided\.  A simple failing case is&#8203;:

$ perl5\.26\.0 \-lwe '\{ package DB; sub f \{ print eval "\\$x" ? "ok" : "fail"; \} \} if\(\+\+\(my $x\)\) \{ DB&#8203;::f\(\); \}'
fail

If you look at the optree\, you'll see that the body of the if\(\) had a
nextstate op at compile time\, but it got nulled out&#8203;:

6     \<;> nextstate\(main 5 \-e&#8203;:1\) v&#8203;:\{ \->7
\-     \<1> null vK/1 \->d
9        \<|> and\(other\->a\) vK/1 \->d
8           \<1> preinc sK/1 \->9
7              \<0> padsv\[$x&#8203;:6\,9\] sPRM/LVINTRO \->8
\-           \<@&#8203;> scope vK \->\-
\-              \<;> ex\-nextstate\(main 8 \-e&#8203;:1\) v \->a
c              \<1> entersub vKS \->d
\-                 \<1> ex\-list K \->c
a                    \<0> pushmark s \->b
\-                    \<1> ex\-rv2cv sK/1 \->\-
b                       \<\#> gv\[\*DB&#8203;::f\] s \->c

This matters because the eval from DB\-land goes looking up the call chain
for the last call from a non\-DB location\, and uses the lexical state
represented by whatever COP was current \(i\.e\.\, most recently executed\)
at the point of that call\.  Without the elision\, the current COP at
the time of the entersub would be the immediately preceding nextstate
op with sequence number 8\, which would mean that $x would be in scope
\(sequence range \(6\, 9\]\)\.  But with that COP elided\, the current COP
is the one preceding the whole if\(\) construct\, with sequence number 5\,
leaving $x out of scope\.

Note that $x is nevertheless in scope for the subroutine call as judged
at compile time of that call\.  You can pass $x to the sub\, for example\,
and that'll be the lexical $x\.  That works because at compile time lexical
scoping is determined on the basis of the COPs as they're generated\,
before elision happens\.  So you can happily pass the lexical $x to a
sub that at runtime doesn't see that $x being in scope at the call site\.

Adding a no\-op statement such as "1;" inside the block\, either before
or after the sub call\, can prevent elision of the nextstate op\, making
the eval behave correctly\.

I said this was a regression in 5\.13\.7\, but that's not entirely correct\.
The test case provided by the bug reporter\, with a while\(\) loop\, starts
failing in 5\.13\.7\, and that can be seen by changing "if\(\+\+\(my $x\)\)" in the
above one\-liner to "while\(\+\+\(my $x\) > $a\+\+\)" or similar\.  But the if\(\)
example fails all the way back to pre\-5\.6 perls\.  It's manifesting by
different routes\, but there's basically a problem that our compilation
code doesn't appreciate the importance of having the correct COP current
at the time of a sub call\.

If the downside of this problem were only evals from the DB package
going wrong\, we might accept that as the price of an optimisation\,
especially since it only applies when not in debug mode\.  \(With \-d on
the command line\, the nextstate ops become dbstate ops\, and are never
elided\, avoiding the problem\.\)  But there are other things controlled
by COPs\.  Runtime warnings seem to be OK&#8203;: changing lexical warning state
prevents COP elision\, afaics\.  But since 5\.9\.4 COPs have stored a shadow
copy of the compile\-time %^H\, accessible by called subroutines through
\(caller\(0\)\)\[10\]\.  It's semantically important for that to be correct\,
but it's not&#8203;:

$ perl5\.26\.0 \-lwe 'sub f \{ print \+\(\(caller\(0\)\)\[10\] || \{\}\)\->\{foo\} || "fail"; \} if\(\+\+\(my $x\)\) \{ BEGIN \{ $^H\{foo\} = "ok"; \} f\(\); \}'
fail
$ perl5\.10\.0 \-lwe 'sub f \{ print \+\(\(caller\(0\)\)\[10\] || \{\}\)\->\{foo\} || "fail"; \} if\(\+\+\(my $x\)\) \{ BEGIN \{ $^H\{foo\} = "ok"; \} f\(\); \}'
fail

Weirdly\, although the while\(\) version of the eval problem didn't occur
prior to 5\.13\.7\, the while\(\) version of the %^H problem \*does\* occur
prior to 5\.13\.7\.  There are multiple nextstate ops in play here\, and
the one that the called sub sees pre\-5\.13\.7 has the lexical variable
but not the lexical hint\.

So generally there's going to have to be less nulling of nextstate ops\,
at least around entersub ops\.

\-zefram
